### PR TITLE
Added request body buffer regeneration

### PR DESCRIPTION
**Pull request by dylandreimerink**
I made a fix for issue #41 and #42.
Optional support for body buffer regeneration.

### CODE REVIEW

File: lib/resty/upload.lua [hunk 1]
```diff
@@ -93,21 +111,37 @@ function _M.set_timeout(self, timeout)
...
 end
 
+local function append_body(self, data_chunk)
+    if self.resore_body then
+        ngx_append_body(data_chunk)
+    end
+end
+
+local function finish_body(self)
+    if self.resore_body then
```
Comment (bungle) [inline lib/resty/upload.lua]: typo here

File: lib/resty/upload.lua [hunk 2]
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
...
+        end
+
         return nil, err
     end
 
     return setmetatable({
         sock = sock,
         size = chunk_size or CHUNK_SIZE,
         line_size = max_line_size or MAX_LINE_SIZE,
+        resore_body = restore_body_buffer or false,
```
Comment (bungle) [inline lib/resty/upload.lua]: `resore` -> `restore`? Also `or false` not needed, though, doesn't hurt much either.

File: lib/resty/upload.lua [hunk 3]
```diff
@@ -145,6 +181,8 @@ local function read_body_part(self)
 
         local data = sock:receive(2)
         if data == "--" then
+            append_body(self, "\r\n--" .. self.boundary .. "--\r\n")
```
Comment (bungle) [inline lib/resty/upload.lua]: We always concatenate strings here. Even when `restore_body` is `false`. Same thing happens in other `append_body` places. Maybe just checking the boolean variable doesn't warrant its own function and that can be inlined. Then we also solve this issue.

File: lib/resty/upload.lua [hunk 4]
```diff
@@ -93,21 +111,37 @@ function _M.set_timeout(self, timeout)
...
+        ngx_finish_body()
+    end
+end
 
 local function discard_line(self)
     local read_line = self.read_line
 
     local line, err = read_line(self.line_size)
     if not line then
+        finish_body(self);
```
Comment (bungle) [inline lib/resty/upload.lua]: semicon `;` not needed.

**Comment by dylandreimerink** [comment]
I fixed the typo and implemented you suggestion. I am not sure though if it is faster to append multiple times like I have done now or to create a new string and append once.

**Comment by bungle** [comment]
Yes, previously you created strings even when not appending anything.

File: lib/resty/upload.lua [hunk 5]
```diff
@@ -93,21 +111,39 @@ function _M.set_timeout(self, timeout)
     return sock:settimeout(timeout)
 end
 
+local function append_body(self, ...)
```
Comment (bungle) [inline lib/resty/upload.lua]: I don't think you need to use varargs here, as you only ever have at most 3 args. Also you don't need to create a table with varargs, you can `select("#", ...)` and `select(1, ...)`.

**Comment by bungle** [comment]
@dylandreimerink thanks for working on this. I think that one remaining thing is to add test(s).

**Comment by dylandreimerink** [comment]
I took a look at the source code for [ngx.req.append_body](https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_req_body.c#L614) which is quite big and does a memcpy. So I think it is best to ditch the whole idea of a separate function ( like u suggested before ) and concatenate a single string in the main functions. Then calling ngx.req.append_body once right before ngx.req.finish_body to minimize overhead.

I will work on that as soon as I have time

**Comment by dylandreimerink** [comment]
This should do it. I replaced the appending by a [string buffer](https://www.lua.org/pil/11.6.html) which should make it more performant. If this passes the only thing left to do is to make tests. I have to admit however that I do not yet understand how the this unit test framework works, so i may need a little bit of guidance.

**Comment by dylandreimerink** [comment]
Working from the assumption that append_body is relatively speaking slow since it has to copy data from the lua vm to nginx the idea is to data in lua for as long as possible. Since lua strings are immutable they have to be recreated and copied every time you want to concatenate to an existing string. The solution would be to have a native implementation of a mutable string or string buffer. But since I have little hope for that in the near future I think the use of a table to store a list of strings and concatenating them once is the best option.

The reason I am going to these lengths is because I will be using this feature in proxy with a high throughput, so for me personally speed and optimization is quite important. So in light of that I will benchmark the different solutions so we can make an informed decision instead of having to work based on assumptions.

**Comment by dylandreimerink** [comment]
I made the [benchmark](https://github.com/dylandreimerink/lua-resty-upload-benchmark). The [results](https://github.com/dylandreimerink/lua-resty-upload-benchmark/blob/master/results.txt) are to me a bit unexpected. I tested 4 variations and the original, the summarized results are as follows:

Name | total time | time per req (avg) | reqests per second(avg)
-- | -- | -- | --
Base test | 87.77 | 8.777 | 11393.38
complex string buffer test | 90.987 | 9.099 | 10990.54
simple string buffer test | 90.971 | 9.097 | 10992.56
string concat test | 90.843 | 9.084 | 11008.04
No buffer test | 90.233 | 9.023 | 11082.48

The variation without the string buffer is the fastest but the difference is almost negligible.
So I went with that solution.

**Comment by dylandreimerink** [comment]
@bungle Have you had time to look at the last changes?

File: lib/resty/upload.lua [hunk 6]
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
...
+        end
+
         return nil, err
     end
 
     local read_line, err = sock:receiveuntil("\r\n")
     if not read_line then
+
+        if restore_body_buffer then
+            ngx_finish_body()
```
Comment (bungle) [inline lib/resty/upload.lua]: The previous read, on line 72, was successful, but we didn't write the request body back, is this intended? Maybe there are many similar cases in code, but no easy way to solve? E.g. on errors we don't write the request body back.

File: lib/resty/upload.lua [hunk 7]
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
...
     local line, err = read_line(self.line_size)
     if not line then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
         return nil, err
     end
 
+    if self.restore_body_buffer then
+        ngx_append_body(line .. "\r\n")
```
Comment (bungle) [inline lib/resty/upload.lua]: I think this is fine, but you could also have:
```lua
ngx_append_body(line)
ngx_append_body("\r\n")
```

File: lib/resty/upload.lua [hunk 8]
```diff
@@ -145,6 +181,10 @@ local function read_body_part(self)
 
         local data = sock:receive(2)
         if data == "--" then
+            if self.restore_body_buffer then
+                ngx_append_body("\r\n--" .. self.boundary .. "--\r\n")
```
Comment (bungle) [inline lib/resty/upload.lua]: Same here:
```lua
ngx_append_body("\r\n--")
ngx_append_body(self.boundary)
ngx_append_body(ngx_append_body("--\r\n"))
```

Not a big deal.

File: lib/resty/upload.lua [hunk 9]
```diff
@@ -161,10 +201,18 @@ local function read_body_part(self)
             end
         end
 
+        if self.restore_body_buffer then
+            ngx_append_body("\r\n--" .. self.boundary .. "\r\n")
```
Comment (bungle) [inline lib/resty/upload.lua]: This coud be cached before the while loop:

```lua
local boundary_delimiter = "\r\n--" .. self.boundary .. "\r\n"
```

And then here:
```lua
ngx_append_body(boundary_delimiter)
```

File: lib/resty/upload.lua [hunk 10]
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
     while true do
         local preamble = read2boundary(size)
         if not preamble then
+            if self.restore_body_buffer then
+                ngx_append_body("--" .. self.boundary)
```
Comment (bungle) [inline lib/resty/upload.lua]: Ditto here:
```lua
ngx_append_body("--")
ngx_append_body(self.boundary)
```

Yes, bikeshedding :-).

File: lib/resty/upload.lua [hunk 11]
```diff
@@ -203,7 +264,10 @@ local function read_header(self)
 end
 
 
-local function eof()
+local function eof(self)
+    if self.restore_body_buffer then
+        ngx_finish_body()
+    end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 12]
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
     while true do
         local preamble = read2boundary(size)
         if not preamble then
+            if self.restore_body_buffer then
+                ngx_append_body("--" .. self.boundary)
+            end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 13]
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
...
             break
+        else
+            if self.restore_body_buffer then
```
Comment (bungle) [inline lib/resty/upload.lua]: You can combine `else` and `if` here to `elseif`.

File: lib/resty/upload.lua [hunk 14]
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
     while true do
         local preamble = read2boundary(size)
         if not preamble then
+            if self.restore_body_buffer then
+                ngx_append_body("--" .. self.boundary)
+            end
             break
+        else
```
Comment (bungle) [inline lib/resty/upload.lua]: new line before this.

File: lib/resty/upload.lua [hunk 15]
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
...
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
         return nil, nil, "line too long: " .. line .. dummy .. "..."
     end
 
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 16]
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
...
 
+    if self.restore_body_buffer then
+        ngx_append_body(line .. "\r\n")
+    end
+
     local dummy, err = read_line(1)
     if dummy then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 17]
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
...
     end
 
+    if self.restore_body_buffer then
+        ngx_append_body(line .. "\r\n")
+    end
+
     local dummy, err = read_line(1)
     if dummy then
+        if self.restore_body_buffer then
+            ngx_finish_body()
```
Comment (bungle) [inline lib/resty/upload.lua]: do we also need to restore that `dummy`?

File: lib/resty/upload.lua [hunk 18]
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
...
 
-
 local function discard_line(self)
     local read_line = self.read_line
 
     local line, err = read_line(self.line_size)
     if not line then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 19]
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
         return nil, err
     end
 
+    if restore_body_buffer then
+        ngx_init_body(chunk_size)
+    end
+
     local read2boundary, err = sock:receiveuntil("--" .. boundary)
     if not read2boundary then
+
```
Comment (bungle) [inline lib/resty/upload.lua]: this new line is not needed.

File: lib/resty/upload.lua [hunk 20]
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
...
+        end
+
         return nil, err
     end
 
     return setmetatable({
         sock = sock,
         size = chunk_size or CHUNK_SIZE,
         line_size = max_line_size or MAX_LINE_SIZE,
+        restore_body_buffer = restore_body_buffer or false,
```
Comment (bungle) [inline lib/resty/upload.lua]: `or false` not strictly needed, as this is only for internal use.

File: lib/resty/upload.lua [hunk 21]
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
     return sock:settimeout(timeout)
 end
 
-
```
Comment (bungle) [inline lib/resty/upload.lua]: restore this new line.

File: lib/resty/upload.lua [hunk 22]
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
...
 
+    if self.restore_body_buffer then
+        ngx_append_body(line .. "\r\n")
+    end
+
     local dummy, err = read_line(1)
     if dummy then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 23]
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
...
         return nil, "line too long: " .. line .. dummy .. "..."
     end
 
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 24]
```diff
@@ -122,6 +152,9 @@ local function discard_rest(self)
     while true do
         local dummy, err = sock:receive(size)
         if err and err ~= 'closed' then
+            if self.restore_body_buffer then
+                ngx_finish_body()
+            end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 25]
```diff
@@ -137,6 +170,9 @@ local function read_body_part(self)
 
     local chunk, err = read2boundary(self.size)
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua [hunk 26]
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
 
     local line, err = read_line(self.line_size)
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
Comment (bungle) [inline lib/resty/upload.lua]: new line after this.

File: lib/resty/upload.lua (hunk 6 above)
Comment (dylandreimerink) [reply lib/resty/upload.lua]: The sock:receiveuntil returns an iterator Lua function and doesn't actually read data from the buffer so there is noting to restore.

File: lib/resty/upload.lua (hunk 17 above)
Comment (dylandreimerink) [reply lib/resty/upload.lua]: The dummy is there to detect if the line we are trying to read exceeded the size of self.line_size. When this happens the request can't be processed further. This means that the remainder of the request will not be restored anyway. So there is no point is appending the one byte.

File: lib/resty/upload.lua (hunk 7 above)
Comment (dylandreimerink) [reply lib/resty/upload.lua]: I don't see any point in not concatenating the string here, the same goes for the rest of the remarks regarding concatenation.

**Pull request closed without merging**